webapi/src/main/scala/org/knora/webapi/messages/admin/responder/projectsmessages/ProjectsMessagesADM.scala

```diff
@@ -505,130 +511,104 @@ case class ProjectADM(
-
-    new ProjectIdentifierADM(
-      maybeIri =
-        sf.validateAndEscapeOptionalProjectIri(maybeIri, throw BadRequestException(s"Invalid user project $maybeIri")),
-      maybeShortname = sf.validateAndEscapeOptionalProjectShortname(
-        maybeShortname,
-        throw BadRequestException(s"Invalid user project shortname $maybeShortname")
-      ),
-      maybeShortcode = sf.validateAndEscapeOptionalProjectShortcode(
-        maybeShortcode,
-        throw BadRequestException(s"Invalid user project shortcode $maybeShortcode")
-      )
-    )
-  }
-}
+sealed trait ProjectIdentifierADM { self =>
+  def asIriOption: Option[String] =
+    self match {
+      case IriIdentifier(value) => Some(value.value)
+      case _                    => None
+    }
 
-/**
- * Represents the project's identifier. It can be an IRI, shortcode or shortname.
- *
- * @param maybeIri       the project's IRI.
- * @param maybeShortname the project's shortname.
- * @param maybeShortcode the project's shortcode
- */
-class ProjectIdentifierADM private (
-  maybeIri: Option[IRI] = None,
-  maybeShortname: Option[String] = None,
-  maybeShortcode: Option[String] = None
-) {
-
-  // squash and return value.
-  val value: String = List(
-    maybeIri,
-    maybeShortname,
-    maybeShortcode
-  ).flatten.head
-
-  def hasType: ProjectIdentifierType =
-    if (maybeIri.isDefined) {
-      ProjectIdentifierType.IRI
-    } else if (maybeShortcode.isDefined) {
-      ProjectIdentifierType.SHORTCODE
-    } else {
-      ProjectIdentifierType.SHORTNAME
+  def asShortcodeOption: Option[String] =
+    self match {
+      case ShortcodeIdentifier(value) => Some(value.value)
+      case _                          => None
     }
 
-  /**
-   * Tries to return the value as an IRI.
-   */
-  def toIri: IRI =
-    maybeIri.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.IRI' type.")
-    )
+  def asShortnameOption: Option[String] =
+    self match {
+      case ShortnameIdentifier(value) => Some(value.value)
+      case _                          => None
+    }
 
-  /**
-   * Returns an optional value of the identifier.
-   */
-  def toIriOption: Option[IRI] =
-    maybeIri
+  def asUuidOption: Option[String] =
+    self match {
+      case UuidIdentifier(value) => Some(UuidIdentifier.makeProjectIri(value.value))
+      case _                     => None
+    }
+}
+
+object ProjectIdentifierADM {
 
   /**
-   * Tries to return the value as an SHORTNAME.
+   * Represents [[IriIdentifier]] identifier.
+   *
+   * @param value that constructs the identifier in the type of [[ProjectIri]] value object.
    */
-  def toShortname: String =
-    maybeShortname.getOrElse(
-      throw DataConversionException(s"Identifier $value is not of the required 'ProjectIdentifierType.SHORTNAME' type.")
-    )
+  final case class IriIdentifier(value: ProjectIri) extends ProjectIdentifierADM
```

Comment:
I don't fully understand why we need an `IriIdentifier`, `ShortcodeIdentifier` etc. if we have `ProjectIri`, `Shortcode` etc. already. To me this looks overengineered. Can't we just use the value objects `ProjectIri`, `Shortcode` etc. directly?

Reply:
Of course this would be the ideal solution, but the design of the responder and SPARQL quarry force to distinguish between identifier we store in the project. It could be improved but that forces even deeper refactoring which wasn't the scope of this task. I went at the beginning for the simplest solution, buy actually reusing getByIri route, but then decided to make it better and at least refactor the IdentifierADM, as an exercise and for a better quality code.